platform/mbed_retarget.cpp

```diff
@@ -904,21 +904,68 @@ extern "C" long PREFIX(_flen)(FILEHANDLE fh)
     return size;
 }
 
-extern "C" char Image$$RW_IRAM1$$ZI$$Limit[];
+#if defined(TARGET_NUMAKER_IOT_M487) || defined(TARGET_NUMAKER_PFM_M2351) || defined(TARGET_NUMAKER_PFM_M453) || defined(TARGET_NUMAKER_PFM_M487) || defined(TARGET_NUMAKER_PFM_NANO130) || defined(TARGET_NUMAKER_PFM_NUC472) || defined(TARGET_GR_LYCHEE) || defined(TARGET_RZ_A1H) || defined(TARGET_RZ_A1XX)
+extern char Image$$ARM_LIB_HEAP$$Base[];
+extern char Image$$ARM_LIB_STACK$$Base[];
+#define HEAP_BASE (Image$$ARM_LIB_HEAP$$Base)
+#define STACK_BASE (Image$$ARM_LIB_STACK$$Base)
+#else
+extern char Image$$RW_IRAM1$$ZI$$Limit[];
+extern char Image$$ARM_LIB_STACK$$ZI$$Base[];
+#define HEAP_BASE (Image$$RW_IRAM1$$ZI$$Limit)
+#define STACK_BASE (Image$$ARM_LIB_STACK$$ZI$$Base)
+#endif
 
-extern "C" MBED_WEAK __value_in_regs struct __initial_stackheap _mbed_user_setup_stackheap(uint32_t R0, uint32_t R1, uint32_t R2, uint32_t R3)
+extern __value_in_regs struct __initial_stackheap _mbed_user_setup_stackheap(uint32_t R0, uint32_t R1, uint32_t R2, uint32_t R3)
 {
-    uint32_t zi_limit = (uint32_t)Image$$RW_IRAM1$$ZI$$Limit;
-    uint32_t sp_limit = __current_sp();
-
-    zi_limit = (zi_limit + 7) & ~0x7;    // ensure zi_limit is 8-byte aligned
-
     struct __initial_stackheap r;
-    r.heap_base = zi_limit;
-    r.heap_limit = sp_limit;
+    r.heap_base = (uint32_t)HEAP_BASE;
+    r.heap_limit = (uint32_t)STACK_BASE;
```

Comment:
Why assign `STACK_BASE` to `r.heap_limit` rather than e.g. `Image$$ARM_LIB_HEAP$$ZI$$Limit`? Heap and stack may arrange non-consecutive or stack in lower memory/heap in higher memory (like Nuvoton targets)?   The same question to `$Sub$$__rt_lib_init`.

Reply:
So yeah it looks strange. The problem is that currently on most targets `ARM_LIB_HEAP` is not defined, but `ARM_LIB_STACK` is. This should be changed in the future. 
Since it is misleading I think I add another macro `HEAP_LIMIT`.